app/models/user.rb

```diff
@@ -120,11 +119,21 @@ def validate_pin_pattern
   # The patron creator service creates a new patron record in the Sierra ILS, and comes back with
   # a success/failure response.
   # Accepts a response from the microservice, and returns.
-  def send_request_to_patron_creator_service
+  def send_request_to_patron_creator_service(pin_code)
+    LogWrapper.log('DEBUG', {
+       'message' => "user.send_request_to_patron_creator_service: start with self=#{self || 'NA'}",
+       'method' => "#{model_name}.send_request_to_patron_creator_service",
+       'status' => "start"
+    })
+
+    if pin_code.blank?
+      pin_code = self.pin
```

Comment:
What are the circumstances where the passed `pin_code` would be different from `self.pin`?
In the call it's clear these are the same:
`FindAvailableUserBarcodeJob.perform_later(user: self, pin_code: self.pin)`
so it's confusing why you need to pass in the pin separately here.

Reply:
Serialization issues.  I'll make it more explicit with a doc string.

Reply:
Wouldn't it be easier to pass user ID and fetch the object from the DB and use the pin so you can remove this check.  Rather than passing it two different ways (`pin_code` and `self.pin`)  and then having to verify which one contains the value. This would less be confusing for someone new to understand in the future.

Reply:
Not necessarily.  We might run into some issues with Devise -- I've seen some funky behavior around when, exactly, the User object gets persisted into the database, and what's in the session and the cache.  Plus, passing the User as a serialized "handler" object is how the hook is meant to be run.  In the future, we might want to replace backends, and I'd like to be as standardized as possible.

Reply:
I'd like to also be able to call this method from anywhere within the app, not just the job.  If I remove this explicit passing and check, my complexity would still be there, just hidden under the covers.  I'd like to make any possible issues be as visible as possible, for future maintenance and upgrades (a newer version of RoR might not have the same issues, and we'll see the "extra" code and clean it).

Reply:
@daryanypl  @thisisstephenbetts  Rails 6 introduces custom serializers for objects which are passed as arguments to ActiveJob.

Reply:
Hi Darya, I'd left this section of code alone until now to not get in the way of the conversation between you and Garvita. But as I know you were keen to get this resolved I took a look late last week and did a little research. 

I agree that 
```FindAvailableUserBarcodeJob.perform_later(user: self, pin_code: self.pin)```
raises a bit of an alert in my mind when reading the code. It makes the developer worry why it's done that way. 

And then the 

```
     if pin_code.blank? 
         pin_code = self.pin
```

raises another alert, which compounds the first. It's unclear why we are anticipating pin_code being blank; and if user was not serialized properly, it's unclear why `self.pin` would work now. Additionally, it raises a concern that the pin and user _could_ get out of sync (if the PIN was updated before the active job executes, for example). 

You said during your talk, in reply to Leonard, that only the user_id is stored in the active_job table in the DB. So if we know that, why not just pass that explicitly as Garvita suggested? It's the simplest info to store, and very easy to retrieve from the db. I don't see why Devise or the cache or the session would get in the way of that if called explicitly.
Then the method of the job could just start

```
def perform(user_id)
    user = User.find(user_id)
```

which would be clear and mean the `if pin_code.blank?` conditional can be removed.

That would also work if it was run from anywhere in the app, though if you wanted to avoid unnecessarily refetching the user from the database  (I don't think it's a big deal to do that, but if you _do_ want to avoid it) you could do a refactor so that the subsequent code in the perform method is separated out (either as a User method or a separate function). E.g., 

```
def perform(user_id)
    user = User.find(user_id)

    user.obtain_barcode_in_sierra
```

or 

```
def perform(user_id)
    user = User.find(user_id)

    obtain_barcode_for_user(user)
```

I don't think that is really necessary, though.

I think this, as Garvita had suggested, would be an improvement. It simplifies the code (has fewer execution paths and permutations of data) and makes it clearer.